Add option to ban former passwords

mfrister opened the pull request:
OpenProject Issue: https://www.openproject.org/issues/778

mfrister committed:
Add option to prevent former passwords from being reused

Squashed from single commits:
* Add checks to prevent former passwords from being reused
* Make number of banned former passwords configurable

  Also group settings into fieldsets on the authentication settings
  page.
* Update TestUnit fixtures with UserPasswords
* Add migration for creating user passwords
* Fix new user password not being immediately used
* Add cuke for ban of former passwords
* Fix admin info showing whether admin password was changed
* Fix password check for user without password and fix tests
* Fix password complexity feature after translation changes
* Adapt my_controller_spec to UserPasswords
* Order UserPasswords by id

  created_at is not unique within one second, this breaks MySQL tests.

  id is guaranteed to be unique and monotonically increasing on MySQL
  and on Postgres(at least as long cache is not set for the sequence).
* Rename secure_compare to secure_equals?
* Immediately use reduced number of banned former passwords

  Only check number of former passwords as defined in settings. The
  number of stored passwords per user is only reduced when a user
  updates the password.
* Add missing copyright notices
* Update changelog

mfrister committed:
Make old migration salting existing user passwords independent of User

Remove code from user, remove test and adapt migration.

ulferts commented on app/models/user.rb:
```diff
@@ -690,6 +683,21 @@ def candidates_for_project_allowance project
   end
 
+  def current_password
+    self.passwords.first
```
You did not specify an order here. Neither did you specify a default scope in UserPassword. So you will receive an arbitrary password (based on the db).

mfrister commented:
We still need to decide what to do with the Apache Subversion authentication module.

See https://www.openproject.org/issues/778

mfrister committed:
Remove broken Apache authentication module

The move of user passwords to the UserPassword table breaks the
authentication module. We should either add a similar module later
which asks the Rails process to do permission checks or move all of this
to a plugin.

cratz commented on app/models/user.rb:
```diff
@@ -692,6 +690,17 @@ def candidates_for_project_allowance project
 
+  def former_passwords_include?(password)
+    return false if Setting[:password_count_former_banned].to_i == 0
+    passwords.any?{ |f| f.same_as_plain_password?(password) }
```
If we 'limit' the passwords to 'Setting[:password_count_former_banned]' than we use the correct number of former passwords for the first password change after lowering  'password_count_former_banned' .

cratz commented on app/models/user_password.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  # constant-time comparison algorithm to prevent timing attacks
+  def self.secure_compare(a, b)
```
This method returns a bool so the name should end with '?'.

mfrister commented:
I made the proposed changes and squashed everything to the three commits, let's see what Travis says.

mfrister commented:
Added copyright notices and changelog. I should start reading the [code review guidelines](https://www.openproject.org/projects/openproject/wiki/Code_Review_Guidelines) I wrote.

mfrister committed:
Fix _authentication.html.erb syntax
